### PR TITLE
Redesign pipelines

### PR DESCRIPTION
**Pull request by SzymonPobiega**
### Behaviors

Behaviors are now singletons (as opposed to instance-per-pipeline-execution). The `Behavior` base class offers new extensions points that allow a behavior to perform any necessary warm-up and cool-down activities in an async way

```
public override Task Warmup()
{
   //Perform warm-up activities 
   return Task.FromResult(default(object));
}

public override Task Cooldown()
{
   //Perform cool-down activities 
   return Task.FromResult(default(object));
}
```

or

```
public override async Task Warmup()
{
   //Perform warm-up activities 
   await PerformWarmup();
}

public override async Task Cooldown()
{
   //Perform cool-down activities 
   await PerformCooldown();
}
```

Behaviors now also have access to the information about the pipeline they are in

```
public override void Invoke(Context context, Action next)
{
    Console.WriteLine(PipelineInfo.Name);
    Console.WriteLine(PipelineInfo.PublicAddress);
    next();
}
```

### Registering a satellite

```
var messageProcessorPipeline = context.AddSatellitePipeline("Timeout Message Processor", "Timeouts");
messageProcessorPipeline.Register<MoveFaultsToErrorQueueBehavior.Registration>();
messageProcessorPipeline.Register<FirstLevelRetriesBehavior.Registration>();
messageProcessorPipeline.Register<TimeoutMessageProcessorBehavior.Registration>();
```

The `TimeoutMessageProcessorBehavior` behavior inherits from a new `SatelliteBehavior` class that makes it easier to migrate from the old `ISatellite`.

### Checking status of features

Previously, enabling and disabling features has been done via setting a bool flag (`true` = `enabled`, `false` = `disabled`). The key was features's class `FullName`. 

The new API uses the same keys but replaces the bool flag with an explicit enumeration (`Enabled`, `Disabled`, `Activated`, `Deactivated`) which allows for querying which feature has actually been activated (enabled features might not get activated if their, or their dependencies', prerequisites are not met).

### Conditional behavior registration

`RegisterStep` now allows to make the behavior registration conditional via `IsEnabled` virtual method

```
public override bool IsEnabled(ReadOnlySettings settings)
{
    var enabled = CheckIfBehaviorShouldBeEnabled();
    return enabled;
}
```

There is also a convenient base class that allows for creating a feature-dependent registrations that only have effect if associated feature has been *activated*.

```
public class Registration : FeatureDependentRegisterStep<Features.FirstLevelRetries>
{
    public Registration()
        : base("FirstLevelRetries", typeof(FirstLevelRetriesBehavior), "Performs first level retries")
    {
        InsertBeforeIfExists("ReceivePerformanceDiagnosticsBehavior");
    }
}
```

## Description

This pull started with changing the way the public receive address is delivered to the behaviors but quickly turned into a redesign of pipelines since there was no way to actually make it work with current approach to the receive behavior.
 * Optionally pass the public address (if provided e.g. by distributor) to the receiver
 * If receiver is not given the public address, he has to provide one and return
 * Public address is passed in the initialization step to all the behaviors
 * All behaviors are not singletons per pipeline with a defined lifecycle (`construct -> init -> start -> stop`)
 * Satellite pipelines are first class citizens on the `FeatureConfigurationContext`. That means that features are not responsible for defining pipelines
 * Removed `ISatellite` and replaced existing satellites with behaviors. The new bahavior lifecycle API is enough to fulfill all the responsibilities of `ISatellite`
 * Refactored the way how feature enabling and activation is signalized so that now it is possible to query for state of a feature (Enabled, Disabled, Activated, Deactivated)
 * Changed the behavior of `FirstLevelRetriesBehavior` so that it only gets added to the pipeline if the feature it is defined in has been activated

### CODE REVIEW

File: src/NServiceBus.Core/DataBus/DataBus.cs [hunk 1]
```diff
@@ -48,8 +48,8 @@ protected internal override void Setup(FeatureConfigurationContext context)
                 context.Container.ConfigureComponent<DefaultDataBusSerializer>(DependencyLifecycle.SingleInstance);
             }
 
-            context.Pipeline.Register<DataBusReceiveBehavior.Registration>();
-            context.Pipeline.Register<DataBusSendBehavior.Registration>();
+            context.PipelinesCollection.Register<DataBusReceiveBehavior.Registration>();
```
Comment (andreasohlund) [inline src/NServiceBus.Core/DataBus/DataBus.cs]: `PipelinesCollections` reads vierd to me. How about context.Pipelines?

File: src/NServiceBus.Core/Features/FeatureActivator.cs [hunk 2]
```diff
@@ -123,12 +123,11 @@ public FeaturesReport SetupFeatures(FeatureConfigurationContext context)
                 }
             }
 
-            settings.PreventChanges();
-
             foreach (var feature in enabledFeatures)
             {
                 ActivateFeature(feature, enabledFeatures, context);
             }
+            settings.PreventChanges();
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Features/FeatureActivator.cs]: Still seems wrong to allow the features to change state when they are activated?

Comment (SzymonPobiega) [inline src/NServiceBus.Core/Features/FeatureActivator.cs]: Features get `ReadOnlySettings` so they can't mess with it unless they downcast.

**Comment by SzymonPobiega** [comment]
@Particular/engineers what do you think about it?

**Comment by johnsimons** [comment]
@SzymonPobiega 
> All behaviors are not singletons per pipeline with a defined lifecycle (construct -> init -> start -> stop)

What do you mean ?

> Refactored the way how feature enabling and activation is signalized so that now it is possible to query for state of a feature (Enabled, Disabled, Activated, Deactivated)

why is there a need for this ?

> Changed the behavior of FirstLevelRetriesBehavior so that it only gets added to the pipeline if the feature it is defined in has been activated

This should be done regardless of the outcome of this PR, should just go ahead and do it in develop

> Satellite pipelines are first class citizens on the FeatureConfigurationContext. That means that features are not responsible for defining pipelines

Not following, a feature can add steps to the pipeline, is that what u mean ?

**Comment by johnsimons** [comment]
@SzymonPobiega how about splitting this PR into more bite size chinks of work, eg:
- Address refactoring
- Pipeline improvements
- Removal of `ISatellite` and conversion to use a pipeline

Would this work?

**Comment by johnsimons** [comment]
@SzymonPobiega are you planing to do all these changes for v6 timeline ?

**Comment by johnsimons** [comment]
@SzymonPobiega it would be nice to provide some examples how the API is going to look, eg:
- How do I migrate from my current `ISatellite` to this new pipeline thing?
- How do I register a step in the main pipeline vs a specific satellite ?

**Comment by SzymonPobiega** [comment]
@johnsimons 

@SzymonPobiega 
>> All behaviors are not singletons per pipeline with a defined lifecycle (construct -> init -> start -> stop)

>What do you mean ?

I mean, there is one instance of a given behavior in a given pipeline. Behavior instances are no longer created for each execution of the pipeline. There are lifecycle methods (`Initialize`, `Starting`, `Stopped`) that a behavior can override to execute some action

>> Refactored the way how feature enabling and activation is signalized so that now it is possible to query for state of a feature (Enabled, Disabled, Activated, Deactivated)

>why is there a need for this ?

Because of the next thing. The registration for the FLR behavior should only register it in the satellite pipeline if the FLR feature has been activated (enabled + prerequisites fulfilled). The previous system with bool flags in settings allowed only to check if a feature was enabled/disabled

>> Changed the behavior of FirstLevelRetriesBehavior so that it only gets added to the pipeline if the feature it is defined in has been activated

>This should be done regardless of the outcome of this PR, should just go ahead and do it in develop

Could not be done without changing the way features activation/deactivation is signalled.

>> Satellite pipelines are first class citizens on the FeatureConfigurationContext. That means that features are not responsible for defining pipelines

> Not following, a feature can add steps to the pipeline, is that what u mean ?

Yes, a feature such as `TimeoutManager` can add pipelines and steps

```
var messageProcessorPipeline = context.AddSatellitePipeline("Timeout Message Processor", "Timeouts");
messageProcessorPipeline.Register<MoveFaultsToErrorQueueBehavior.Registration>();
messageProcessorPipeline.Register<FirstLevelRetriesBehavior.Registration>();
messageProcessorPipeline.Register<TimeoutMessageProcessorBehavior.Registration>();
```

**Comment by SzymonPobiega** [comment]
@johnsimons the address refactorings are moved to a separate branch. The only thing left here is the satellites and changing the lifecycle of behaviors.

**Comment by johnsimons** [comment]
> I mean, there is one instance of a given behavior in a given pipeline. Behavior instances are no longer created for each execution of the pipeline. There are lifecycle methods (Initialize, Starting, Stopped) that a behavior can override to execute some action

Why do we need this, wouldn't it be better to just create for each execution ?

**Comment by johnsimons** [comment]
> The registration for the FLR behavior should only register it in the satellite pipeline if the FLR feature has been activated (enabled + prerequisites fulfilled)

I am confused, what is the difficulty of doing this ?
This is how we currently do things, on the Setup method of a feature we register behaviors, that seems to work fine ?

**Comment by johnsimons** [comment]
```c#
var messageProcessorPipeline = context.AddSatellitePipeline("Timeout Message Processor", "Timeouts");
messageProcessorPipeline.Register<MoveFaultsToErrorQueueBehavior.Registration>();
messageProcessorPipeline.Register<FirstLevelRetriesBehavior.Registration>();
messageProcessorPipeline.Register<TimeoutMessageProcessorBehavior.Registration>();
```

Interesting code @SzymonPobiega, can we set a call to review it ? I like it.

**Comment by andreasohlund** [comment]
>Why do we need this, wouldn't it be better to just create for each
execution ?

I think we need to get on a call to align on this since this is coming from
some deep discussions with @danielmarbach and others

@szymonpobiega can you schedule it?

**Comment by danielmarbach** [comment]
@SzymonPobiega I tried to replace you in a meeting today. I think I did a good job to convince @johnsimons 

To be honest, I think we should not use `OnStarting` / `OnStopping`. On Methods are used as event invocators in .NET.

How about `Warmup` and `Cooldown`?

I like that we have explicit methods. This allows us to introduce async there as well in the future

**Comment by SzymonPobiega** [comment]
@danielmarbach I love `Warmup` and `Cooldown`. Thanks!

**Comment by SzymonPobiega** [comment]
@danielmarbach Done. I hope I haven't messed up stuff horribly with asyncs

**Comment by SzymonPobiega** [comment]
@johnsimons do you have any comments for this PR after the Monday call? What's left before I can merge in you opinion?

**Comment by johnsimons** [comment]
@SzymonPobiega I think the PR is good, however I think it would be beneficial to update the description with usage examples (think of it as the doco).
This way we all have a good feeling for what the API usage will be and then can provide good feedback on naming and flow of the API in general, does that sound reasonable ?

**Comment by SzymonPobiega** [comment]
@Particular/engineers can you review the new/changed APIs?

**Comment by indualagarsamy** [comment]
@SzymonPobiega - is there an associated doco PR that I can look at? I apologize, if its mentioned here, and I just missed it.

**Comment by SzymonPobiega** [comment]
@indualagarsamy not yet, for now only the stuff in the description of the PR

**Comment by danielmarbach** [comment]
@SzymonPobiega People usually use Task.FromResult(true) or Task.FromResult(1) to something constant and not null

File: src/NServiceBus.Core/Pipeline/Behavior.cs [hunk 3]
```diff
@@ -44,5 +74,30 @@ public void Invoke(TContext context, Action<TContext> next)
...
+        {
+            PipelineInfo = pipelineInfo;
+        }
+
+        /// <summary>
+        /// Allows a behavior to perform any necessary warm-up activities (such as priming a cache), possibly in an async way.
+        /// </summary>
+        public virtual Task Warmup()
+        {
+            return Task.FromResult(default(object));
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Pipeline/Behavior.cs]: Matter of taste but I usually use Task.FromResult(0), or (1) or (true)

File: src/NServiceBus.Core/Pipeline/Behavior.cs [hunk 4]
```diff
@@ -44,5 +74,30 @@ public void Invoke(TContext context, Action<TContext> next)
...
+        }
+
+        /// <summary>
+        /// Allows a behavior to perform any necessary cool-down activities, possibly in an async way.
+        /// </summary>
+        public virtual Task Cooldown()
+        {
+            return Task.FromResult(default(object));
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Pipeline/Behavior.cs]: Matter of taste but I usually use Task.FromResult(0), or (1) or (true)

File: src/NServiceBus.Core/Pipeline/StageConnector.cs [hunk 5]
```diff
@@ -12,12 +14,41 @@ public abstract class StageConnector<TFrom, TTo> :IStageConnector, IBehavior<TFr
...
+        {
+            PipelineInfo = pipelineInfo;
+        }
+
+        /// <summary>
+        /// Allows a behavior to perform any necessary warm-up activities (such as priming a cache), possibly in an async way.
+        /// </summary>
+        public virtual Task Warmup()
+        {
+            return Task.FromResult(default(object));
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Pipeline/StageConnector.cs]: Matter of taste but I usually use Task.FromResult(0), or (1) or (true)

File: src/NServiceBus.Core/Pipeline/StageConnector.cs [hunk 6]
```diff
@@ -12,12 +14,41 @@ public abstract class StageConnector<TFrom, TTo> :IStageConnector, IBehavior<TFr
...
+        }
+
+        /// <summary>
+        /// Allows a behavior to perform any necessary cool-down activities, possibly in an async way.
+        /// </summary>
+        public virtual Task Cooldown()
+        {
+            return Task.FromResult(default(object));
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Pipeline/StageConnector.cs]: Matter of taste but I usually use Task.FromResult(0), or (1) or (true)

File: src/NServiceBus.Core/Timeout/Hosting/Windows/TimeoutDispatcherProcessorBehavior.cs [hunk 7]
```diff
@@ -44,21 +34,26 @@ public bool Handle(TransportMessage message)
...
         {
             TimeoutPersisterReceiver.Start();
+            return Task.FromResult(default(object));
         }
 
-        public void Stop()
+        public override Task Cooldown()
         {
             TimeoutPersisterReceiver.Stop();
+            return Task.FromResult(default(object));
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Timeout/Hosting/Windows/TimeoutDispatcherProcessorBehavior.cs]: Matter of taste but I usually use Task.FromResult(0), or (1) or (true)

File: src/NServiceBus.Core/Timeout/Hosting/Windows/TimeoutDispatcherProcessorBehavior.cs [hunk 8]
```diff
@@ -44,21 +34,26 @@ public bool Handle(TransportMessage message)
             return true;
         }
 
-        public void Start()
+        public override Task Warmup()
         {
             TimeoutPersisterReceiver.Start();
+            return Task.FromResult(default(object));
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Timeout/Hosting/Windows/TimeoutDispatcherProcessorBehavior.cs]: Matter of taste but I usually use Task.FromResult(0), or (1) or (true)

File: src/NServiceBus.Core/Unicast/Behaviors/TransportReceiveToPhysicalMessageProcessingConnector.cs [hunk 9]
```diff
@@ -1,18 +1,53 @@
...
+    class TransportReceiveToPhysicalMessageProcessingConnector : StageConnector<TransportReceiveContext, PhysicalMessageProcessingStageBehavior.Context>, IDisposable
     {
+        IPerformanceCounterInstance successRateCounter;
+        IPerformanceCounterInstance failureRateCounter;
+
+        public override Task Warmup()
+        {
+            successRateCounter = PerformanceCounterHelper.TryToInstantiatePerformanceCounter("# of msgs successfully processed / sec", PipelineInfo.PublicAddress);
+            failureRateCounter = PerformanceCounterHelper.TryToInstantiatePerformanceCounter("# of msgs failures / sec", PipelineInfo.PublicAddress);
+            return Task.FromResult(default(object));
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/Behaviors/TransportReceiveToPhysicalMessageProcessingConnector.cs]: Matter of taste but I usually use Task.FromResult(0), or (1) or (true)

File: src/NServiceBus.Core/Unicast/Transport/ReceivePerformanceDiagnosticsBehavior.cs [hunk 10]
```diff
@@ -1,14 +1,32 @@
...
-    class ReceivePerformanceDiagnosticsBehavior : PhysicalMessageProcessingStageBehavior
+    [SkipWeaving]
+    class ReceivePerformanceDiagnosticsBehavior : PhysicalMessageProcessingStageBehavior, IDisposable
     {
+        IPerformanceCounterInstance counter;
+
+        public override Task Warmup()
+        {
+            counter = PerformanceCounterHelper.TryToInstantiatePerformanceCounter("# of msgs pulled from the input queue /sec", PipelineInfo.PublicAddress);
+            return Task.FromResult(default(object));
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/Transport/ReceivePerformanceDiagnosticsBehavior.cs]: Matter of taste but I usually use Task.FromResult(0), or (1) or (true)

File: src/NServiceBus.Core/Unicast/Transport/TransportReceiver.cs [hunk 11]
```diff
@@ -102,11 +93,11 @@ public virtual void Start()
...
             }
 
-            InitializePerformanceCounters(dequeueSettings.QueueName);
-
             Logger.DebugFormat("Pipeline {0} is starting receiver for queue {0}.", Id, dequeueSettings.QueueName);
 
-            receiver.Init(dequeueSettings);
+            var dequeueInfo = receiver.Init(dequeueSettings);
+            pipeline.Initialize(new PipelineInfo(id, dequeueInfo.PublicAddress));
+            pipeline.Warmup();
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/Transport/TransportReceiver.cs]: I'd say push the task as high up in the call stack as possible. So make `pipeline.Warmup();` return a task as well and synchronize here or even push it up more

File: src/NServiceBus.Core/Unicast/Transport/TransportReceiver.cs [hunk 12]
```diff
@@ -142,6 +128,8 @@ protected virtual void InnerStop()
             }
 
             receiver.Stop();
+            pipeline.Cooldown();
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/Transport/TransportReceiver.cs]: I'd say push the task as high up in the call stack as possible. So make pipeline.Cooldown(); return a task as well and synchronize here or even push it up more

**Comment by SzymonPobiega** [comment]
@danielmarbach done with `Cooldown`. Also, I changed the noop tasks to use `true` instead of `null`. @andreasohlund do you still like it after going async?

**Comment by danielmarbach** [comment]
@SzymonPobiega I like but my vote doesn't count because I paired on it with you ;)

You should also do a rebase

**Comment by andreasohlund** [comment]
I'm good with the async, +1 for going ahead

**Pull request merged**